oak_functions/loader/src/server.rs

```diff
@@ -149,25 +148,63 @@ impl FixedSizeBodyPadder for Response {
     }
 }
 
+/// Trait for implementing extensions, to implement new native functionality.
+pub trait OakApiNativeExtension {
+    /// Similar to `invoke_index` in [`wasmi::Externals`], but may return a result to be
+    /// written into the memory of the `WasmState`.
+    fn invoke(
+        &self,
+        wasm_state: &WasmState,
+        args: wasmi::RuntimeArgs,
+    ) -> Result<Result<Option<ExtensionResult>, OakStatus>, wasmi::Trap>;
+
+    /// Metadata about this Extension, including a unique internal index, the exported host function
+    /// name, and the function's signature
+    fn get_metadata(&self) -> ExtensionMetadata;
+}
+
+pub type BoxedExtension = Box<dyn OakApiNativeExtension + Send + Sync>;
+
+/// Metadata about an [`OakApiNativeExtension`].
+pub struct ExtensionMetadata {
+    /// Name of the exported host function.
+    pub name: String,
+    /// Unique internal index corresponding to the exported host function. It is required when
+    /// invoking the host function.
+    pub index: usize,
+    /// Signature of the exported host function.
+    pub signature: wasmi::Signature,
+}
+
+// Result of the invocation of an extension. The [`WasmState`] will write the `bytes` into the
+// memory location specified by `buf_ptr_ptr` and `buf_len_ptr`.
+pub struct ExtensionResult {
+    pub bytes: Vec<u8>,
+    pub buf_ptr_ptr: AbiPointer,
+    pub buf_len_ptr: AbiPointer,
+}
+
 /// `WasmState` holds runtime values for a particular execution instance of Wasm, handling a
 /// single user request. The methods here correspond to the ABI host functions that allow the Wasm
 /// module to exchange the request and the response with the Oak functions server. These functions
 /// translate values between Wasm linear memory and Rust types.
-struct WasmState {
+pub struct WasmState {
     request_bytes: Vec<u8>,
     response_bytes: Vec<u8>,
     lookup_data: Arc<LookupData>,
-    #[allow(dead_code)]
-    tf_model: Arc<Option<TensorFlowModel>>,
     instance: Option<wasmi::ModuleRef>,
     memory: Option<wasmi::MemoryRef>,
     logger: Logger,
     metrics_proxy: Option<PrivateMetricsProxy>,
+    /// A mapping of internal host functions to the corresponding [`OakApiNativeExtension`].
+    extensions_indices: Arc<HashMap<usize, Arc<BoxedExtension>>>,
```

Comment:
We create a new instance of `WasmState` for each incoming request. `Arc` is needed, because `HashMap` is not clonable. Using `&HashMap` would require specifying lifetimes. I think `Arc` is cleaner and is more consistent with `lookup_data` which is `Arc`.